KAFKA-3831: Prepare for updating new-consumer-based Mirror Maker's default partition assignment strategy to round robin

vahidhashemian opened the pull request:
This patch adds proper warning message and necessary doc updates for updating the default partition assignment strategy of Mirror Maker from range to round robin. The actual switch would occur as part of a major release cycle (to be scheduled).

ijuma commented on core/src/main/scala/kafka/tools/MirrorMaker.scala:
```diff
@@ -83,6 +89,11 @@ object MirrorMaker extends Logging with KafkaMetricsGroup {
   def main(args: Array[String]) {
 
     info("Starting mirror maker")
+    println("WARNING: The default partition assignment strategy of mirror maker will change from 'range' to 'roundrobin' " +
```
We should probably only warn if the partition assignment strategy is not the one that will become the default. This way people have a way to get rid of the warning.

vahidhashemian replied:
Thanks @ijuma. I'm just thinking about a perhaps rare or unlikely scenario in which, the user has currently set their partition assignment strategy to `roundrobin`. Later after the major release (when default is changed) they decide to go back to `range` assuming that is still the default (so they remove the property from their consumer config). This way, they don't see any warning (as the warning will supposedly be removed when the default is actually changed) and will keep using the `roundrobin` assignment even though they think they're using `range`.

Very unlikely scenario, but just wanted to make sure it is still ok to go with your suggestion. Thanks.

ijuma replied:
I did think about that and thought it was OK. But let's see what others say.

hachikuji replied:
Before committing to this, I'm wondering what the plan is for compatibility. I doubt users care much whether MM uses range or round-robin, but they would probably expect that they can safely do a rolling upgrade. So a couple points:
1. What does the old consumer do if assignment strategies differ? If we can't make the change in a compatible way, perhaps we should just leave the default as range. We're ultimately planning to deprecate the old consumer anyway.
2. It's actually possible with the new consumer to make this change in a compatible way by enabling both round-robin and range assignment. The coordinator ensures that an assignment strategy which all consumers support is chosen. The downside is that it increases the overhead of the rebalance protocol (basically the subscriptions need to be sent to the coordinator twice). It might make sense to recommend doing two rolling bounces in that case: the first one enables both assignors, the second enables only round-robin. On the other hand, there probably aren't many users who have as many topics as LinkedIn, so perhaps the extra overhead is not significant and it may be reasonable to turn on both assignors by default?

vahidhashemian replied:
@hachikuji I've been looking more closely into MM today to better understand your comments. I have to admit MM is not one of the best and most up-to-date documented features of Kafka (for example, I could not find anything concrete about how to make MM produce an identical cluster with same partitions and messages).

But from what I understand so far,
1. Ideally, MM should run continuously. If it stops and resumes there is a chance for duplicate / lost messages.
2. I don't see a way for consumers in MM to follow different config; e.g. use different assignment strategy. All consumers follow the same strategy (the default or what's given in consumer config). In other words, each MM instance runs with a single partition assignment strategy.
3. In a running MM consumers cannot come and go, and so MM uses a fixed config throughout execution.

With my understanding above I'm having a hard time realizing the compatibility issue, and why switching from one default strategy to another could be problematic compatibility-wise. That's why I'm guessing I've got something wrong above (for example, my second bullet above seems to be in conflict with your statement about enabling both strategies). In either case I'd appreciate your clarification.

hachikuji replied:
@vahidhashemian MM runs as a distributed system. So to upgrade configuration without downtime, you need to do a rolling restart of the MM processes. This means that some instances in the cluster will be on the old config and some will be on the new config. We could tell users that they have to shutdown all of their MM processes, perform the upgrade, and then restart, but that's not too great, right?

vahidhashemian replied:
@hachikuji Thanks for the explanation. I agree that stopping and restarting everything for an upgrade is not ideal. I'm thinking about scenarios with multiple MM processes (several `kafka.tools.MirrorMaker` instances) and I can think of when we would want to mirror a subset of topics in each MM process, with the most granular being one topic per MM process (which wouldn't really be affected by this change). So I'm guessing one or both of the following is possible (though I'm not sure how):
- a MM process mirroring some partitions of a topic (and not all of its partitions)
- multiple MM processes mirror the same message log (in parallel)

Is that so? Or I'm missing something? Thanks.

hachikuji replied:
@vahidhashemian Haha, I'm not sure I see the source of the confusion. The MM processes are basically just a consumer group. They all subscribe to the same set of topics and they depend on the assignment protocol to balance the partitions from those topics among the group. So this requires that they all have a compatible assignment process, right?

vahidhashemian replied:
@hachikuji Yes, I understand that MM works as a consumer group consuming from a number of given topics. So my confusion seems to be on trivial matters. Am I correct to assume that once MM starts we can easily stop some of the consumers and restart them? I now feel that this is what you meant by MM processes (consumers rather than the MM command). Thanks :)

vahidhashemian replied:
Or perhaps what makes more sense now is that one could have multiple MM processes that are associated with the same consumer group.

hachikuji replied:
@vahidhashemian To clarify a little. When you start the MM process (by executing the MM command), one consumer instance will be created. Other MM processes (potentially running on different hosts) will have their own consumer instance. When multiple MM processes share the same consumer group Id, they will work in concert as a single consumer group in order to balance the load. The problem is that they must all agree on the assignment protocol. In a rolling upgrade scenario, you will have some MM processes running on some hosts which still use the old config, and you will have some MM processes running on other hosts which use the newer config. So what's the behavior when some MM hosts are supporting 'range' and some are supporting 'round-robin'? In general, it just doesn't work, but there is an option available for the new consumer (as I described above). Does that make sense?

vahidhashemian replied:
@hachikuji Yes, thanks. That's what I realized when I guess you were writing your last post :)
I'm going to look further into that last piece, which is how it would work in the new consumer, as you explained. I might bug you again for clarification. Thanks for bearing with me no this.

vahidhashemian replied:
@hachikuji Quick question: In your first comment above what do you mean by enabling (or turning on) both assignors? Aren't both range and round robin assignors already enabled for the new consumer to use? Thanks.

hachikuji replied:
@vahidhashemian Ah yeah, what I mean is that you can actually specify both assignors in the consumer's configuration (partition.assignment.strategy accepts a list). The names of all assignment strategies will get forwarded to the coordinator, which will attempt to choose one which is shared by all members in the group. The idea was to use this in a rolling upgrade. For example, say you have consumers A and B which are using range assignment. You can change the configuration of consumer A to be "round-robin,range" and do a restart. The coordinator will continue to choose "range" since it's the only strategy both consumers support. You can then change the configuration of consumer B to also be "round-robin,range" and then the coordinator will choose "round-robin" (preference is implied by the order in the list). Ideally, you would then do another round of rolling restarts in which you remove "range" from the list.

Sounds great, right? The problem is that supporting multiple assignment strategies carries some overhead in the rebalance protocol. We end up sending the topic subscription list once for each supported strategy. In some MM use cases (LI in particular), this can be a problem since the consumer might be subscribed to thousands of topics. Still, it might be a reasonable tradeoff since most users won't have so many topics and it provides an easier upgrade path.

vahidhashemian replied:
@hachikuji Thanks a lot for the explanation. What you suggested makes sense (I had missed the fact that multiple assignors can be provided in the consumer config). If I'm not mistaken the overhead you mentioned starts [here](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java#L137). I think your suggestion is a proper way to go about the rolling upgrade (assuming changing the default assignor is imminent).

vahidhashemian commented:
@ewencp @hachikuji This PR follows your advice on [KAFKA-3818](https://issues.apache.org/jira/browse/KAFKA-3818) to prep for the actual change in a later release. @ijuma kindly reviewed this a while ago, and I would appreciate your feedback as well. Thanks.

vahidhashemian commented:
@hachikuji I updated the warning message based on the earlier discussion. Please let me know if you see any issues, and also your take on @ijuma's [comment](https://github.com/apache/kafka/pull/1499#discussion_r66838550) on the previous commit.

hachikuji commented:
@vahidhashemian I'm still a little unsure about the right way to go here. The problem with my suggestion of switching to "round-robin,range" and finally just to "round-robin" is that the user won't be able to skip releases (which we have tried to allow up to this point). In that case, maybe we're better off just jumping straight to "round-robin" only. Perhaps we could make the switch to round-robin directly, but give users an "--upgrade" flag which enables both range and round-robin. What do you think?

vahidhashemian commented:
@hachikuji Would it be possible to keep the "roundrobin,range" assignment configuration for a few releases so chances of skipping releases causing issues with MM are reduced? Or we don't want to keep it around for long because of the overhead that comes with specifying multiple assignment?

hachikuji commented:
@vahidhashemian Yeah, the extra overhead makes me a little uncomfortable. Maybe the warning message for this release should simply inform users that "round-robin" will usually provide better load balancing and that it will become the default in the future. Then we can decide later whether we want to provide a nicer upgrade path. What do you think?

hachikuji commented:
The other thing I was thinking about is whether we ought to provide an "--assignment-strategy" option to make it easier for users to change.

vahidhashemian commented:
@hachikuji That's fine with me. I could change the warning message back to the previous version.
Introducing an `--assignment-strategy` option sounds like a great idea too. I can open a JIRA for this improvement. Sounds good?

hachikuji commented:
Yes, that sounds fine to me. Not sure if @ijuma has any thoughts.

hachikuji commented on core/src/main/scala/kafka/tools/MirrorMaker.scala:
```diff
@@ -60,6 +60,11 @@ import scala.util.control.ControlThrowable
  *       3. Mirror Maker Setting:
  *            abort.on.send.failure=true
+ *
+ *      The default partition assignment strategy of the new-consumer-based mirror maker will change from 'range'
```
This comment seems like overkill given the message below and the fact that we have a JIRA to track this change.

vahidhashemian replied:
Sure, no problem. I can remove this comment.

hachikuji commented on core/src/main/scala/kafka/tools/MirrorMaker.scala:
```diff
@@ -184,6 +189,12 @@ object MirrorMaker extends Logging with KafkaMetricsGroup {
+
+        println("WARNING: The default partition assignment strategy of the new-consumer-based mirror maker will " +
+          "change from 'range' to 'roundrobin' in the future (so that a better load balancing can be achieved). " +
+          "One easy way to keep using your current default strategy ('range') is to add the following to the new " +
```
I feel it would probably be more helpful here to describe how the user can change to 'round-robin' in advance of the upcoming change (there really shouldn't be any need for users to stick with 'range'). Also, maybe we should only print this message if the user has not overridden the assignment strategy?

vahidhashemian replied:
So instead of advising users how to stick with `range` we advise how they can switch to `round-robin` in advance. I'll change the message. I'll also look into how we can be smart about when to print it, as you suggested. Thanks.

hachikuji reviewed (commented):
@vahidhashemian Left a couple minor comments. I'm considering pulling this into 0.10.1.

vahidhashemian committed:
KAFKA-3831: Prepare for updating new-consumer-based Mirror Maker's default partition assignment strategy to round robin

This patch adds proper warning message for updating the default partition assignment strategy of Mirror Maker from range to round robin. The actual switch would occur as part of a major release cycle (to be scheduled.

vahidhashemian commented:
@hachikuji I made the changes you mentioned your last review. I hope I've not missed anything. Thanks.

hachikuji commented on core/src/main/scala/kafka/tools/MirrorMaker.scala:
```diff
@@ -43,6 +43,7 @@ import org.apache.kafka.common.record.Record
 import scala.collection.JavaConversions._
 import scala.collection.mutable.HashMap
 import scala.util.control.ControlThrowable
+import org.apache.kafka.clients.consumer.{ ConsumerConfig => NewConsumerConfig }
```
nit: are the extra spaces conventional?

vahidhashemian replied:
I'll remove the extra spaces, just checked some other files and realized what format is being used. Thanks.

hachikuji commented on core/src/main/scala/kafka/tools/MirrorMaker.scala:
```diff
@@ -184,6 +185,13 @@ object MirrorMaker extends Logging with KafkaMetricsGroup {
         }
+
+        if (!Utils.loadProps(options.valueOf(consumerConfigOpt)).keySet().contains(NewConsumerConfig.PARTITION_ASSIGNMENT_STRATEGY_CONFIG))
+          println("WARNING: The default partition assignment strategy of the new-consumer-based mirror maker will " +
```
Maybe we should print to stderr instead? This is what we do in other tools for deprecation warnings (see e.g. ConsoleConsumer).

vahidhashemian replied:
Sure, I'll do that.

hachikuji commented on core/src/main/scala/kafka/tools/MirrorMaker.scala:
```diff
@@ -184,6 +185,13 @@ object MirrorMaker extends Logging with KafkaMetricsGroup {
           System.exit(1)
         }
+
+        if (!Utils.loadProps(options.valueOf(consumerConfigOpt)).keySet().contains(NewConsumerConfig.PARTITION_ASSIGNMENT_STRATEGY_CONFIG))
```
nit: This line seems a little wasteful. Seems like we could store the loaded properties in a variable and pass it down to `createOldConsumers` and `createNewConsumers` directly instead of the path.

vahidhashemian replied:
Sure, I was contemplating that myself. I'll update in the next patch.

hachikuji reviewed (commented):
Few minor comments, but looks good overall.

hachikuji commented on core/src/main/scala/kafka/tools/MirrorMaker.scala:
```diff
@@ -184,6 +187,13 @@ object MirrorMaker extends Logging with KafkaMetricsGroup {
+            "change from 'range' to 'roundrobin' in an upcoming release (so that a better load balancing can be achieved). " +
+            "If you prefer to make this switch in advance of that release add the following to the corresponding new-" +
+            "consumer config:\n" +
+            "* partition.assignment.strategy=org.apache.kafka.clients.consumer.RoundRobinAssignor")
```
Is the asterisk intentional?

vahidhashemian replied:
I meant to use it as a bullet. I could change that if you think it's confusing.

hachikuji replied:
It looks a little weird when I see the output on the command line. Maybe just keep it on the same line and surround it with quotes? Not a big deal though.

vahidhashemian replied:
Sure, np. I'll update this one along with the other typo.

vahidhashemian replied:
The only thing is I didn't add line breaks in the message. If the config line is appended to the line before there is a chance it breaks in two lines depending on the terminal size.

hachikuji replied:
Either way seems fine to me. I think users are probably used to copying text that wraps.

vahidhashemian replied:
Fair enough. Then I'll append it. Thanks.

hachikuji commented on core/src/main/scala/kafka/tools/MirrorMaker.scala:
```diff
@@ -184,6 +187,13 @@ object MirrorMaker extends Logging with KafkaMetricsGroup {
+
+        if (!consumerProps.keySet().contains(NewConsumerConfig.PARTITION_ASSIGNMENT_STRATEGY_CONFIG))
+          System.err.println("WARNING: The default partition assignment strategy of the new-consumer-based mirror maker will " +
+            "change from 'range' to 'roundrobin' in an upcoming release (so that a better load balancing can be achieved). " +
```
nit: do we need the "a" before "better"?

vahidhashemian replied:
Hmm, probably not. I'll get rid of it.

hachikuji commented:
LGTM. Thanks for the patch!
